File: Windows/Services/FactorGovernanceService.cs
namespace CytisusTrading.Windows;

public sealed record FactorReviewResult(
    IReadOnlyList<FactorItem> Factors,
    IReadOnlyList<string> ChangedFactorIds);

public interface IFactorGovernanceService
{
    FactorReviewResult RunReview(
        IReadOnlyList<FactorItem> factors,
        int reviewCount,
        double maxFactorWeight);
}

public sealed class FixtureFactorGovernanceService : IFactorGovernanceService
{
    public FactorReviewResult RunReview(
        IReadOnlyList<FactorItem> factors,
        int reviewCount,
        double maxFactorWeight)
    {
        var reviewed = factors.Select(factor => factor.Clone()).ToList();
        var changed = new List<string>();

        foreach (var factor in reviewed)
        {
            switch (factor.FactorId)
            {
                case "volatility-term-structure":
                    factor.EvidenceWindows += 1;
                    if (factor.EvidenceWindows >= 6 && reviewCount >= 2)
                    {
                        factor.State = FactorState.Active;
                        factor.Weight = Math.Min(0.05, maxFactorWeight);
                        factor.Reason = "Passed twice; enable with a 5% cap";
                    }
                    else
                    {
                        factor.Reason = "First pass complete; awaiting confirmation";
                    }
                    changed.Add(factor.FactorId);
                    break;

                case "volume-impact":
                    factor.Ic -= 0.003;
                    if (reviewCount >= 2)
                    {
                        factor.State = FactorState.Retired;
                        factor.Weight = 0;
                        factor.Reason = "Failed three reviews; 126-day cooldown";
                    }
                    else
                    {
                        factor.Reason = "Second failure; probation continues";
                    }
                    changed.Add(factor.FactorId);
                    break;

                case "short-term-reversal":
                    factor.EvidenceWindows += 1;
                    factor.State = FactorState.Shadow;
                    factor.Reason = "Candidate data gate passed; moved to shadow review";
                    changed.Add(factor.FactorId);
                    break;
            }
        }

        return new FactorReviewResult(reviewed, changed);
    }
}

File: Sources/Services/FactorGovernanceService.swift
import Foundation

struct FactorReviewResult {
    let factors: [FactorItem]
    let changedFactorIDs: [String]
}

protocol FactorGovernanceServicing {
    func runReview(
        factors: [FactorItem],
        reviewCount: Int,
        maxFactorWeight: Double
    ) -> FactorReviewResult
}

struct FixtureFactorGovernanceService: FactorGovernanceServicing {
    func runReview(
        factors: [FactorItem],
        reviewCount: Int,
        maxFactorWeight: Double
    ) -> FactorReviewResult {
        var reviewed = factors
        var changed: [String] = []

        for index in reviewed.indices {
            switch reviewed[index].factorID {
            case "volatility-term-structure":
                reviewed[index].evidenceWindows += 1
                if reviewed[index].evidenceWindows >= 6 && reviewCount >= 2 {
                    reviewed[index].state = .active
                    reviewed[index].weight = min(0.05, maxFactorWeight)
                    reviewed[index].reason = "Passed twice; enable with a 5% cap"
                } else {
                    reviewed[index].reason = "First pass complete; awaiting confirmation"
                }
                changed.append(reviewed[index].factorID)

            case "volume-impact":
                reviewed[index].ic -= 0.003
                if reviewCount >= 2 {
                    reviewed[index].state = .retired
                    reviewed[index].weight = 0
                    reviewed[index].reason = "Failed three reviews; 126-day cooldown"
                } else {
                    reviewed[index].reason = "Second failure; probation continues"
                }
                changed.append(reviewed[index].factorID)

            case "short-term-reversal":
                reviewed[index].evidenceWindows += 1
                reviewed[index].state = .shadow
                reviewed[index].reason = "Candidate data gate passed; moved to shadow review"
                changed.append(reviewed[index].factorID)

            default:
                break
            }
        }

        return FactorReviewResult(factors: reviewed, changedFactorIDs: changed)
    }
}
